p2p: update inbound eviction protection for multiple networks, add I2P peers

jonatack opened the pull request:
Continuing the work in #20197 and #20685, this pull updates and abstracts our inbound eviction protection to make it fully ratio-based and easily extensible to peers connected via high-latency privacy networks that we newly support, like I2P and perhaps others soon, as these peers are disadvantaged by the latency criteria of our eviction logic.

It then adds eviction protection for peers connected over I2P.  As described in https://github.com/bitcoin/bitcoin/pull/20685#issuecomment-767486499, we've observed over the past few months that I2P peers have a min ping latency similar to or greater than that of onion peers.

The algorithm is a basically a multi-pass knapsack:

- Count the number of eviction candidates in each of the disadvantaged
  privacy networks.

- Sort the networks from lower to higher candidate counts, so that
  a network with fewer candidates will have the first opportunity
  for any unused slots remaining from the previous iteration.  In
  the case of a tie in candidate counts, priority is given by array
  member order from first to last, guesstimated to favor more unusual
  networks.

- Iterate through the networks in this order.  On each iteration,
  allocate each network an equal number of protected slots targeting
  a total number of candidates to protect, provided any slots remain
  in the knapsack.

- Protect the candidates in that network having the longest uptime,
  if any in that network are present.

- Continue iterating as long as we have non-allocated slots
  remaining and candidates available to protect.

The goal of this logic is to favorise the diversity of our peer connections.

The individual commit messages describe each change in more detail.

Special thank you to Vasil Dimov for the excellent review feedback and the algorithm improvement that made this change much better than it would have been otherwise. Thanks also to Antoine Riard, whose review feedback nudged this change to protect disadvantaged networks having fewer, rather than more, eviction candidates.

vasild commented:
Concept ACK

jonatack commented:
Will update now that #20197 (on which this is based) and #20685 adding I2P support are both merged.

jonatack committed:
p2p: extend inbound eviction protection by network to I2P peers

This commit extends our inbound eviction protection to I2P peers to
favorise the diversity of peer connections, as peers connected
through the I2P network are otherwise disadvantaged by our eviction
criteria for their higher latency (higher min ping times) relative
to IPv4 and IPv6 peers, as well as relative to Tor onion peers.

The `networks` array is order-dependent in the case of a tie in
candidate counts between networks (earlier array members receive
priority in the case of a tie).

Therefore, we place I2P candidates before localhost and onion ones
in terms of opportunity to recover unused remaining protected slots
from the previous iteration, guesstimating that most nodes allowing
both onion and I2P inbounds will have more onion peers, followed by
localhost, then I2P, as I2P support is only being added in the
upcoming v22.0 release.

jonatack committed:
p2p, refactor: rm redundant erase_size calculation in SelectNodeToEvict()

as EraseLastKElements() called in the next line performs the same operation.
Thanks to Martin Zumsande (lightlike) for seeing this while reviewing.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

jonatack committed:
test: speed up and simplify peer_eviction_test

This speeds up the test significantly, which helps when
running it repeatedly.

Suggest reviewing the diff with:

colorMoved = dimmed-zebra
colorMovedWs = allow-indentation-change

jonatack committed:
p2p: update ProtectEvictionCandidatesByRatio() doxygen docs

jonatack committed:
p2p: make ProtectEvictionCandidatesByRatio() fully ratio-based

with a more abstract framework to allow easily extending inbound
eviction protection to peers connected through new higher-latency
networks that are disadvantaged by our inbound eviction criteria,
such as I2P and perhaps other BIP155 networks in the future like
CJDNS.  This is a change in behavior.

The algorithm is a basically a multi-pass knapsack:

- Count the number of eviction candidates in each of the disadvantaged
  privacy networks.

- Sort the networks from lower to higher candidate counts, so that
  a network with fewer candidates will have the first opportunity
  for any unused slots remaining from the previous iteration.  In
  the case of a tie in candidate counts, priority is given by array
  member order from first to last, guesstimated to favor more unusual
  networks.

- Iterate through the networks in this order.  On each iteration,
  allocate each network an equal number of protected slots targeting
  a total number of candidates to protect, provided any slots remain
  in the knapsack.

- Protect the candidates in that network having the longest uptime,
  if any in that network are present.

- Continue iterating as long as we have non-allocated slots
  remaining and candidates available to protect.

Localhost peers are treated as a network like Tor or I2P by aliasing
them to an unused Network enumerator: Network::NET_MAX.

The goal is to favorise diversity of our inbound connections.

Credit to Vasil Dimov for improving the algorithm from single-pass
to multi-pass to better allocate unused protection slots.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

jonatack committed:
test: add combined onion/localhost eviction protection coverage

jonatack committed:
test: add tests for inbound eviction protection of I2P peers

jonatack committed:
test: add combined I2P/onion/localhost eviction protection tests

mzumsande commented:
Concept ACK, will review more closely soon.

vasild reviewed (commented):
Approach ACK 831b3294d92b23f85617e77ea3f90082de73ce3a

jonatack commented on src/net.cpp:
```diff
@@ -896,36 +902,64 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& vEvict
+    if (networks) {
+        // Obtain the minimum number of peers per network to potentially protect.
+        // The base ratio is 1/2 (for the case of peers from only 1 network).
+        const size_t protect_size_per_network{total_protect_size / (2 * networks)};
```
Those are initial (not necessarily final) percentages iff onion, i2p and localhost peers are all present in the eviction candidates. For instance, if no i2p or localhost peers are among the candidats, then the full 25% is made available to onions. And if localhost or i2p peers are present, any initially reserved slots that they don't use are made available for onions. (See the commit messages for a more detailed description). I'm not against adjusting the ratios though.

jonatack replied:
I played around with a more generous version that would protect more than 1/4 in the case of eviction candidates from multiple disadvantaged networks.  Didn't keep it for now, but memoizing the commit: e5c27dd.

jonatack commented on src/net.cpp:
```diff
@@ -896,36 +902,64 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& vEvict
-                           [](const NodeEvictionCandidate& n) { return n.m_is_onion; });
+    // Count number of nodes connected via localhost, I2P, and onion.
+    // Higher positions in the array recover unused slots left by lower ones.
+    std::array<size_t, 3> network_counts{{}};
```
It's used to determine the number of networks from which we have disadvantaged candidates (indeed that could be done with bools also, replacing count_if with a sum of each boolean) but I anticipate possibly also using the actual counts to determine the distribution size or order of the protected slots.

jonatack commented on src/net.cpp:
```diff
@@ -1022,7 +1055,7 @@ bool CConnman::AttemptToEvictConnection()
                                                peer_relay_txes, peer_filter_not_null, node->nKeyedNetGroup,
                                                node->m_prefer_evict, node->addr.IsLocal(),
-                                               node->m_inbound_onion};
+                                               node->m_inbound_onion, node->addr.IsI2P()};
```
> What about passing `enum Network` instead?

Good idea, looks like we can do that easily and it will include future network additions:

```diff
-                                               node->m_inbound_onion, node->addr.IsI2P()};
+                                               node->ConnectedThroughNetwork()};
```

jonatack replied:
Requires a lot of changes but probably still worth it. Will look at replacing `m_is_local` as well, maybe giving it a `Network` value of `NET_UNROUTABLE` for the eviction logic.

jonatack replied:
Passing `enum Network`, instead of multiple booleans, indeed allowed rewriting this to be better abstracted and generalizable. WIP, for now the initial rewrite is in an additional commit.

laanwj commented:
Concept ACK

jonatack commented:
Current status is almost-out-of-WIP after a rewrite and several iterations, just need to finish the latest iteration and reorganize it a bit.

jonatack committed:
p2p: add CompareNodeNetworkTime() comparator struct

to compare and sort peer eviction candidates by the
passed-in is_local (localhost status) and network
arguments, and by longest uptime.

jonatack committed:
test: add ALL_NETWORKS to test utilities

vasild reviewed (commented):
Approach ACK 4890e2f2fdefaa18b0e96bcaf3691fb3f0213572

Some squashing is warranted.

jonatack commented:
> Approach ACK [4890e2f](https://github.com/bitcoin/bitcoin/commit/4890e2f2fdefaa18b0e96bcaf3691fb3f0213572)
> 
> Some squashing is warranted.

Thanks! Indeed. Re-arranging.

ariard commented on src/net.cpp:
```diff
@@ -851,6 +851,12 @@ static bool CompareOnionTimeConnected(const NodeEvictionCandidate& a, const Node
+    if (a.m_is_i2p != b.m_is_i2p) return b.m_is_i2p;
+    return a.nTimeConnected > b.nTimeConnected;
+}
+
```
What would be the purpose of the follow-up you're envisioning ?

AFAIU this current patchset, any unused protected slots across a high-latency network is reallocated to the next one by summing up `protect_size_per_network` and passing through `total_to_protect` ?

jonatack replied:
You're right. The commit message was out of date. Updated.

ariard reviewed (commented):
Approach ACK 831b329

jonatack committed:
test: remove combined onion/localhost eviction protection tests

as we are about the change the behavior sufficiently that when we
have multiple disadvantaged networks and a small number of peers
under test, the number of protected peers per network can be different.

jonatack commented on src/net.cpp:
```diff
@@ -902,26 +902,44 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& vEvict
+            if (network_counts[1]) {
+                // Protect onion peers (connected via our onion service), sorted by longest uptime.
+                const size_t onion_peers_to_protect{total_to_protect - (initial_size - vEvictionCandidates.size())};
+                EraseLastKElements(vEvictionCandidates, CompareOnionTimeConnected, onion_peers_to_protect,
```
Since the inital number of protected slots per network goes down quickly as the number of networks increases, it seems good to not waste any unused slots in this basic greedy knapsack. So we calculate `peers_to_protect` for each iteration having potential peers to protect, in order to recover any unused slots from the previous iteration.

jonatack replied:
Hm, `s/total_to_protect/available_to_protect/`?

duncandean commented:
Concept ACK. Will take a closer look later today :)

jonatack commented:
Thanks for the concept ACKs @mzumsande, @duncandean and @laanwj and the approach ACKs @vasild and @ariard. Now ready for final review. 

This screenshot illustrates the higher latency of I2P peers, which disadvantages them under our inbound eviction criteria even relative to onion peers.

![Screenshot from 2021-06-07 08-56-56](https://user-images.githubusercontent.com/2415484/120973346-544b3400-c75e-11eb-901e-667f36ebcd54.png)

jonatack commented on src/net.cpp:
```diff
@@ -894,40 +893,64 @@ static void EraseLastKElements(
+    const auto networks_size = std::count_if(networks.begin(), networks.end(), [](const Net& n) { return n.count; });
+
+    if (networks_size) {
+        // Obtain the minimum number of peers per network to potentially protect.
+        // The base ratio is 1/2 (for the case of peers from only 1 network).
+        const size_t protect_size_per_network{total_protect_size / (2 * networks_size)};
```
Unused slots can be recovered, so the number protected can be higher, and yes, if there are fewer peers of that network, the number protected can be lower. Perhaps s/minimum/initial/.

vasild replied:
Right, it can be higher or lower. Neither one of "minimum" or "maximum" is suitable. `s/minimum/initial/` looks good.

jonatack replied:
Thanks. This line is now no longer used.

jonatack commented on src/net.cpp:
```diff
@@ -883,6 +871,17 @@ static bool CompareNodeBlockRelayOnlyTime(const NodeEvictionCandidate &a, const
+    CompareNodeNetworkTime(Network n, Network l) : network(n), localhost(l) {}
+    bool operator()(const NodeEvictionCandidate& a, const NodeEvictionCandidate& b) const
+    {
+        if (network == localhost && a.m_is_local != b.m_is_local) return b.m_is_local;
+        if ((a.m_network == network) != (b.m_network == network)) return b.m_network == network;
```
The suggestion is equivalent/does the same thing, assuming the following change is also made to the caller. 
```diff
-                EraseLastKElements(candidates, CompareNodeNetworkTime(n.id, localhost), peers_to_protect,
+                EraseLastKElements(candidates, CompareNodeNetworkTime(n.id, n.id == localhost), peers_to_protect,
```

jonatack replied:
Updated to pass an `is_local` bool.

jonatack commented on src/net.cpp:
```diff
@@ -894,40 +893,64 @@ static void EraseLastKElements(
+    if (networks_size) {
+        // Obtain the minimum number of peers per network to potentially protect.
+        // The base ratio is 1/2 (for the case of peers from only 1 network).
+        const size_t protect_size_per_network{total_protect_size / (2 * networks_size)};
+
+        if (protect_size_per_network) {
+            size_t available_to_protect{protect_size_per_network};
+
+            // Stable-sort networks by ascending candidate count, to allow networks that
+            // have more peers to recover slots not used by networks with fewer peers.
+            std::stable_sort(networks.begin(), networks.end());
+
+            for (const Net& n : networks) {
+                if (n.count == 0) continue;
+
+                // Protect peers of this network, sorted by longest uptime.
+                const size_t peers_to_protect{available_to_protect - (initial_size - candidates.size())};
+                EraseLastKElements(candidates, CompareNodeNetworkTime(n.id, localhost), peers_to_protect,
+                                   [&n](const NodeEvictionCandidate& c) {
+                                       return n.id == localhost ? c.m_is_local : c.m_network == n.id;
+                                   });
+                available_to_protect += protect_size_per_network;
+            }
+        }
```
I think the outcome of Case 2 is ideal.  Case 1 protects many I2P peers because that network gets the unused slots from the localhost allocation.  What change would you suggest?  Keep in mind that this is a subcomponent seen in isolation; another 1/4 of the remaining peers will be protected based on their uptime.

jonatack replied:
Hm, maybe we can tally the number of networks that have more than `protect_size_per_network` peers and try to share the leftover slots between them so that Case 1 protects 1 / 11 / 12 instead of 1 / 15 / 8.

vasild replied:
>  What change would you suggest?

```cpp
void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& candidates)
{
    // Treat localhost connections as a separate network.
    static constexpr Network NET_LOCALHOST{NET_MAX};

    const std::array<Network, 3> networks{NET_LOCALHOST, NET_I2P, NET_ONION};

    // Protect the half of the remaining nodes which have been connected the longest.
    // This replicates the non-eviction implicit behavior, and precludes attacks that start later.
    // To favorise the diversity of our peer connections, reserve up to half of
    // these protected spots for onion, localhost and I2P peers, even if they're
    // not longest uptime overall. This helps protect these higher-latency peers
    // that tend to be otherwise disadvantaged under our eviction criteria.
    const size_t total_to_protect{candidates.size() / 2};

    // Protect up to 25% of `candidates` due to network. We may protect less if
    // there are no candidates that belong to `networks`.

    const size_t max_protect_due_to_network{total_to_protect / 2};
    size_t num_protected{0};

    while (num_protected < max_protect_due_to_network) {
        const size_t remaining_to_protect{max_protect_due_to_network - num_protected};
        const size_t protect_per_network{std::max(remaining_to_protect / networks.size(), 1UL)};
        bool protected_at_least_one{false};

        for (const Network net : networks) {
            const size_t before = candidates.size();
            EraseLastKElements(candidates,
                               CompareNodeNetworkTime(net, net == NET_LOCALHOST),
                               protect_per_network,
                               [&net](const NodeEvictionCandidate& c) {
                                   return net == NET_LOCALHOST ? c.m_is_local : c.m_network == net;
                               });
            const size_t after = candidates.size();
            if (after < before) {
                protected_at_least_one = true;
                num_protected += before - after;
                if (num_protected >= max_protect_due_to_network) {
                    break;
                }
            }
        }

        if (!protected_at_least_one) {
            break;
        }
    }

    // Protect another 25% to 50% of `candidates` due to uptime/longevity. For example, if
    // 25% were protected due to network then we will protect 25% due to uptime, if
    // 20% were protected due to network then we will protect 30% due to uptime.
    const size_t remaining{total_to_protect - num_protected};
    EraseLastKElements(candidates, ReverseCompareNodeTimeConnected, remaining);
}
```

will result in:
```
(number of peers that belong to localhost, i2p, tor => protected peers from localhost, i2p, tor)
1, 20, 20 => 1, 12, 12
0, 20, 20 => 0, 13, 12
```

Feel free to ignore.

jonatack replied:
This seems like a nice approach, multi-pass with simpler data structures. Will adapt the tests to it to check all the test cases.

jonatack replied:
It looks like we should keep the `Net` struct and `<` sort operator so the result can independent of the order of members in the `networks` array, so a hybrid approach, also adding an `is_local` member to the struct to no longer need `net == localhost` throughout the code.

vasild replied:
Yes, with the above snippet, the result may depend on the order of elements in the `networks` array, but only up to 1 which I think it ok - after all, if we want to split evenly e.g. 10 slots among 3 networks, some network will get 4 slots and the rest will get 3 slots.

The current code also depends on the order in `networks` because it does "stable" sort, right?

jonatack replied:
The results make more sense when they depend on number of peers by network, e.g. if you have 10 I2P peers and 3 onion ones, best to give first chance at unused slots to the I2P peers so we protect, say, 3 I2P and 2 onions rather than the opposite.  Stable sort only makes a difference if the peer counts are equal. I'll propose this separately in the last commit for feedback.

vasild replied:
So, if we have 10 I2P peers and 3 onion ones and we want to protect 6 in total, we will protect 3 I2P and 3 onion, this is desired, right?

If we have again 10 I2P and 3 onion and want to protect 5, do we want to protect 3 I2P and 2 onion or 2 I2P and 3 onion? I am not sure. One may argue that protecting one more (but we are talking about just one here) from the rare network (onion) is better because it is easier to find peers from the more popular network (I2P), so the ones from the rare network are more scarce and precious. Or a counter argument could be to keep the proportion - if we have more I2P, then keep more I2P.

Personally, I think it is not so important because it is just 1 peer we are talking about here and I would favor code simplicity over enforcing either approach with more complex code.

jonatack replied:
Yes, this is about tradeoffs. The simplest code was when only a `Network` was passed around all the code without the `is_local` bool everywhere, but that simplicity had a logic encapsulation drawback (perhaps some others), so I rewrote everything and brought back `is_local`. 

I've redone this pull several times because I'd like it to do the right thing, whether networks with more peers get first priority at unused slots or networks with fewer.  If people prefer to have the array order decide independently of the number of peers, it seems we should clearly document that dependency.

I've separated it out to a separate commit, along with the test changes. Am going through everything before pushing the update.

jonatack replied:
My thought is that having separate protection buckets for each of the disadvantaged networks is the essential step here to improving the diversity of our inbound peer connections...and that unused protection slots should be distributed among disadvantaged networks having remaining eviction candidates by descending order of the number of peers.

jonatack committed:
p2p, refactor: improve constness in ProtectEvictionCandidatesByRatio()

jonatack committed:
p2p, refactor: rename vEvictionCandidates to eviction_candidates

in ProtectEvictionCandidatesByRatio()
per current style guide in doc/developer-notes.md

jonatack committed:
p2p: add m_network to NodeEvictionCandidate struct

jonatack committed:
p2p: remove unused m_is_onion member from NodeEvictionCandidate struct

jonatack committed:
p2p: remove unused CompareOnionTimeConnected()

jonatack committed:
p2p: remove unused CompareLocalHostTimeConnected()

jonatack commented:
Updated per the discussions above and rebased.  The CI failure seems unrelated.  I'm running a node on this branch with `-maxconnections=36`.

vasild reviewed (approved):
ACK 42530190d82ee1692a546a8daae89e3ba0f5a3a2

I think the following commits should be squashed into one:
```
e157ed1478 p2p: make ProtectEvictionCandidatesByRatio() fully ratio-based
029256f7ed p2p: give networks with more peers priority for unused protection slots
42530190d8 p2p, refactor: replace NET_LOCALHOST with a bool struct member
```
It could be tedious and time consuming for reviewers to review newly added code that is rehauled in a subsequent commit and then changed again in yet another commit.

jonatack commented on src/test/net_peer_eviction_tests.cpp:
```diff
@@ -36,7 +32,7 @@ std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(const int n_c
             /* prefer_evict */ random_context.randbool(),
             /* m_is_local */ random_context.randbool(),
-            /* m_is_onion */ random_context.randbool(),
+            /* m_network */ static_cast<Network>(random_context.randrange(static_cast<int>(Network::NET_INTERNAL))),
```
Good point, looks like your suggestion with this might do it:
```diff
diff --git a/src/test/net_peer_eviction_tests.cpp b/src/test/net_peer_eviction_tests.cpp
index ca3f943123..f827230e1a 100644
--- a/src/test/net_peer_eviction_tests.cpp
+++ b/src/test/net_peer_eviction_tests.cpp
@@ -4,6 +4,7 @@
 
 #include <netaddress.h>
 #include <net.h>
+#include <test/util/net.h>
 #include <test/util/setup_common.h>
 
 #include <boost/test/unit_test.hpp>
@@ -32,7 +33,7 @@ std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(const int n_c
             /* nKeyedNetGroup */ random_context.randrange(100),
             /* prefer_evict */ random_context.randbool(),
             /* m_is_local */ random_context.randbool(),
-            /* m_network */ static_cast<Network>(random_context.randrange(static_cast<int>(Network::NET_INTERNAL))),
+            /* m_network */ ALL_NETWORKS[random_context.randrange(ALL_NETWORKS.size())],
         });
     }
     return candidates;
diff --git a/src/test/util/net.h b/src/test/util/net.h
index 4e7c446af1..5f3f412cc9 100644
--- a/src/test/util/net.h
+++ b/src/test/util/net.h
@@ -68,7 +68,7 @@ constexpr ConnectionType ALL_CONNECTION_TYPES[]{
     ConnectionType::ADDR_FETCH,
 };
 
-constexpr Network ALL_NETWORKS[]{
+constexpr std::array ALL_NETWORKS{
     Network::NET_UNROUTABLE,
     Network::NET_IPV4,
     Network::NET_IPV6,
```

jonatack replied:
Done.

jonatack replied:
Huh, the CI is barfing on this change. Strange.

MarcoFalke replied:
This might be a compiler bug. I used `static constexpr auto ALL_FEE_ESTIMATE_HORIZONS = std::array{` to work around that.

jonatack replied:
Thanks! Will try that.

jonatack replied:
It worked 👍

ariard commented on src/net.cpp:
```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
+    }
+    // Sort `networks` by descending candidate count, to give networks having more candidates
+    // the first opportunity to recover unused protection slots from the previous iteration.
+    std::stable_sort(networks.begin(), networks.end(), [](Net a, Net b) { return a.count > b.count; });
```
I think I would favor sorting network by ascending candidate count, otherwise I believe it would bias towards anonymity networks with abundant peers, decreasing our local diversity of anonymity network ?

I think this point has been already discussed in this [thread](https://github.com/bitcoin/bitcoin/pull/21261#discussion_r647319655). I agree it's minor.

jonatack replied:
Done. As the unit tests show, one peer can make a difference when the numbers are low, e.g. when -maxconnections has been reduced or only one candidate will be protected.

ariard commented on src/net.cpp:
```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
+            if (before > after) {
+                protected_at_least_one = true;
+                num_protected += before - after;
+                if (num_protected >= max_protect_by_network) break;
```
Following diff doesn't break `net_peer_eviction_tests`, is this an expected behavior ?

```
@@ -943,7 +943,7 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& candid
             if (before > after) {
                 protected_at_least_one = true;
                 num_protected += before - after;
-                if (num_protected >= max_protect_by_network) break;
+                if (num_protected == max_protect_by_network) break;
             }
         }
         if (!protected_at_least_one) break;
 ```

jonatack replied:
Agree, the tests pass with this change.
```diff
-                if (num_protected >= max_protect_by_network) {
+                assert(num_protected <= max_protect_by_network);
+                if (num_protected == max_protect_by_network) {
```
I think it's expected behavior based on these lines:
```cpp
    while (num_protected < max_protect_by_network) {
        const size_t disadvantaged_to_protect{max_protect_by_network - num_protected};
        const size_t protect_per_network{std::max(disadvantaged_to_protect / networks.size(), static_cast<size_t>(1))};
```

vasild replied:
Given that `num_protected` is incremented with a non-constant number, that could be larger than `1`, I assume it could become larger than `max_protect_by_network` at some point. Thus `>=` for comparison instead of `==`.

How could replacing `>=` with `==` and an assert change anything? I think if it changes anything then it must be a trigger of the assert - if it was equal before, then behavior is unchanged (`>=` is the same as `==`) and if it was strictly larger, before we would have executed `break;` and after the change - triggered the assert.

To me that assert looks scary, even if some (non-obvious) arithmetic a few lines above mandates it will never be triggered. The code may be altered in the future which may change this. Also, even if it happens that `num_protected > max_protect_by_network`, then nothing really bad has happened - we protected some peers (more than we intended), the program can continue executing just fine.

_(the code at the time I write this does not contain an assert (a457f34e5039b75ee015b273028c3ee153656d5c))_

jonatack replied:
Right, the assert was just for sanity testing (I wrote more tests but didn't find a way to hit the assert), no plan to add it.

ariard commented on src/net.cpp:
```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
+    while (num_protected < max_protect_by_network) {
+        const size_t remaining_to_protect{max_protect_by_network - num_protected};
+        const size_t protect_per_network{std::max(remaining_to_protect / networks.size(), static_cast<size_t>(1))};
+        bool protected_at_least_one{false};
```
What do you think about commenting the purpose of this variable ?

"If not more eviction candidates remain to protect among disadvantaged networks, ensure we exit cleanly"

jonatack replied:
Done.

ariard reviewed (commented):
I think this is pretty mature, if you can just have a look around this [comment](https://github.com/bitcoin/bitcoin/pull/21261#discussion_r650190229) ? Other comments are less substantial. Once done, I want to check again correctness of the multi-pass algorithm.

jonatack commented on src/net.cpp:
```diff
@@ -894,40 +902,62 @@ static void EraseLastKElements(
+            if (before > after) {
+                protected_at_least_one = true;
+                num_protected += before - after;
+                if (num_protected >= max_protect_by_network) break;
```
Brackets and newlines added

jonatack commented on src/net.cpp:
```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
 }
 
-void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& vEvictionCandidates)
+void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& candidates)
```
done

jonatack commented on src/net.cpp:
```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
+    // Networks to protect: localhost, onion, I2P. In case of equal counts, earlier array members
+    // have the first opportunity to recover unused protection slots from the previous iteration.
+    struct Net { Network id; size_t count; bool is_local; };
+    std::array<Net, 3> networks{{{/* localhost */ NET_MAX, 0, 1}, {NET_ONION, 0, 0}, {NET_I2P, 0, 0}}};
```
done

jonatack commented on src/net.cpp:
```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
+    const size_t initial_size = candidates.size();
+    const size_t total_protect_size{initial_size / 2};
+
+    // Protect up to 25% of the candidates by disadvantaged network.
```
Yes, if there are at least 4 eviction candidates. (Changed per your review feedback from the disadvantaged network with the most candidates to the one with the fewest.)

jonatack commented on src/net.cpp:
```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
+    size_t num_protected{0};
+
+    while (num_protected < max_protect_by_network) {
+        const size_t remaining_to_protect{max_protect_by_network - num_protected};
```
done

jonatack commented:
Thanks for the feedback @vasild and @ariard. All the feedback taken, will re-push after reworking the tests.

jonatack commented:
Updated, taking all review feedback by @vasild and @ariard, and also improving readability and updating the tests.

`git diff 4253019 ce211cd --color-words`

vasild reviewed (approved):
ACK a457f34e5039b75ee015b273028c3ee153656d5c

jonatack commented on src/net.cpp:
```diff
@@ -894,40 +903,71 @@ static void EraseLastKElements(
+    // Disadvantaged networks to protect: I2P, localhost, Tor/onion. In case of equal counts, earlier
+    // array members have first opportunity to recover unused slots from the previous iteration.
+    struct Net { bool is_local; Network id; size_t count; };
+    std::array<Net, 3> networks{{{0, NET_I2P, 0}, {/* localhost */ 1, NET_MAX, 0}, {0, NET_ONION, 0}}};
```
Yeah, I did this to keep the line length down as it's a standard conversion, but sure.

jonatack replied:
Done.

laanwj commented:
I've been running this PR (various versions) on a busy node for the last weeks. Unfortunately this node has a really high `maxconnections` (500, while average number of connected peers is 200) so I just realize I haven't been actually testing the eviction behavior :blush: 

In any case
Code review (and lightly tested) ACK a457f34e5039b75ee015b273028c3ee153656d5c

jonatack commented:
Wow! How long does it take to get up to 200 peers? Mine is restarted a lot but never gets above 35-40.

jonatack commented:
Updated per `git diff a457f34 1b1088d --color-words` for the review suggestion and to add 4 test cases for behavior with 2 and 3 disadvantaged networks and only 4, 7, and 8 eviction candidates.

jonatack commented:
Thanks @vasild and @laanwj.  Updated per git diff a457f34 1b1088d --color-words for the review suggestion and to add 4 test cases, and updated the PR description to thank/credit @vasild and @ariard.

vasild reviewed (approved):
ACK 1b1088d52fbff8b1c9438d6aa8c6edcbdd471457

laanwj commented:
> Wow! How long does it take to get up to 200 peers? Mine is restarted a lot but never gets above 35-40.

Last restart was last week, it gets connections quite quickly. Then again it is connected to all possible networks and has had a >99% uptime for a few years. Either that helps or maybe it's just all spy nodes :slightly_smiling_face:
